### PR TITLE
[feature] adding a naive check for CPU speed

### PR DESCRIPTION
**Pull request by akichidis**
Following discussions this PR is introducing some naive CPU speed checks. We have seen instances of validators where they had features enabled, like power saving modes etc, that prevented them from leveraging the full potential of their CPU. The checks of this PR is stressing the CPU (a single core) via solving two problems:
* finding a fibonacci number
* calculating the factorial

the time is measured and compared against an average expected time. So far I have run this on three different machines (my Mac M2 Max, Mac Studio M1 Ultra & a private-testnet validator) so I need to calibrate further the timing expectations , but so far when running the script I get the following timings:

**Mac M2 Max:**
```
Running Test 1...
Test 1: average time taken: 6.010905 seconds

Running Test 2...
Test 2: average time taken: 0.084257 seconds
```

**Mac Studio M1 Ultra**
```
Running Test 1...
Test 1: average time taken: 6.652985s seconds

Running Test 2...
Test 2: average time taken: 0.093008 seconds
```

**ams-ptn-val-00**
```
Running Test 1...
Test 1: average time taken: 6.192840 seconds

Running Test 2...
Test 2: average time taken: 0.091907 seconds
```

When I put the `Mac M2 Max` to a power saving mode (underclocking the CPU) then I immediately saw an increase on the timing:

**Mac M2 Max (with power saving mode enabled - where CPU is underclocked)**
```
Running Test 1...
Test 1: average time is higher than expected average 7.916298s > 6.700000s

Running Test 2...
Test 2: average time is higher than expected average 0.111570s > 0.100000s
```

### CODE REVIEW

File: src/checks.py [hunk 1]
```diff
@@ -98,6 +98,16 @@ def check_num_cpus() -> Tuple[bool, str, str]:
   output = run_command(["cat /proc/cpuinfo | grep processor | wc -l"])
   return (True, output, None) if int(output) >= MINIMUM_CPU_THREADS else (False, output, "sui-node requires >= 48 CPU threads")
 
+def check_cpu_speed() -> Tuple[bool, str, str]:
+  output = run_command(["./check_cpu_speed"], "lib/bin")
```
Comment (mystenmark) [inline src/checks.py:102]: can we have the program print out the results and then do the comparisons in python? will make it easier to update limits, plus we'll eventually probably want to gather them all in one place.

File: src/lib/src/check_cpu_speed.c [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        return false;
+    } else {
+        printf("Test 2: average time taken: %f seconds\n", average_time);
+        return true;
+    }
+}
+
+int main() {
+    bool test_1_passed = benchmark_fibonacci(10);
+    bool test_2_passed = benchmark_factorial(20);
```
Comment (mystenmark) [inline src/lib/src/check_cpu_speed.c:70]: we should pass in the iteration count from the commandline, or use some other method to ensure that the compiler doesn't evaluate these functions at compile time.

**Review by mystenmark — APPROVED** [review]
As discussed earlier, consider whether we might want to use sysbench, or even just call `head -1000000b /dev/urandom | gzip > /dev/null` and measure how long it takes.

File: src/lib/src/check_cpu_speed.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    double total_time = 0.0;
+    for (int i = 0; i < iterations; i++) {
+
+        // since a child process is created using the clock() won't be accurate anymore as clock() measures the CPU
+        // spent time by the program it self, so any actual spent time from the child process will not be taken into account.
+        // Thus we switch to using the actual wall time instead for this benchmark.
+        struct timespec start, end;
+        clock_gettime(CLOCK_MONOTONIC_RAW, &start);
+
+        int status = system("head -1000000 /dev/urandom | gzip > /dev/null");
```
Comment (mystenmark) [inline src/lib/src/check_cpu_speed.c:40]: you need to do `head -100000b` to specify bytes. without the `b` it reads 1000000 lines which is not a fixed amount of data (depends on where newline characters happen to appear).

**Review by mystenmark — APPROVED** [review]

File: src/lib/src/check_cpu_speed.c (hunk 3 above)
Comment (akichidis) [reply src/lib/src/check_cpu_speed.c:40]: Thanks @mystenmark for the comment. When I tried with the `b` in the end it wasn't working for me
```
head: illegal line count -- 100000b
``` 
and the examples that I saw on the internet they were stating that using the `head -number` would work equally - but now that I am looking this further it seems wrong.

By the head manual it seems that I need to use the `-c` option to count bytes. Indeed when counting the produced bytes I get:

```
$ head  -c 1000000 /dev/urandom | wc -c | awk '{ printf "%.2f MB\n", $1/1024/1024 }'
0.95 MB
```

whereas with
```
$ head  -1000000 /dev/urandom | wc -c | awk '{ printf "%.2f MB\n", $1/1024/1024 }' 
244.19 MB
```

which as you said is simply wrong for fulfilling the command expectation.

I'll update the command to use:
```
head  -c 268435456 /dev/urandom | gzip > /dev/null
```

to test compressing `256 MB` worth of data

**Pull request merged**